### PR TITLE
Fix CGAL v4.7 compatibility

### PR DESCRIPTION
**Pull request by metux**
The special handlings for cgal v4.7 are required since v4.7.0,
instead of v4.7.1.

Signed-off-by: Enrico Weigelt, metux IT consult <info@metux.net>

### CODE REVIEW

**Comment by mhugo** [comment]
Seems correct, thanks !

File: src/algorithm/Intersection3D.cpp [hunk 1]
```diff
@@ -192,7 +192,7 @@ void _intersection_solid_triangle( const MarkedPolyhedron& pa, const CGAL::Trian
     // triangle decomposition
     std::list<MarkedPolyhedron> decomposition;
     Is_not_marked criterion;
-#if CGAL_VERSION_NR < 1040701000 // version 4.7
+#if CGAL_VERSION_NR < 1040700000 // version 4.7
```
Comment (lrineau) [inline src/algorithm/Intersection3D.cpp]: `1040701000` is the `CGAL_VERSION_NR` of CGAL-4.7.0. The `1000` is the build number, set to `1000` for public releases.

See [the documentation about `CGAL_VERSION_NR`](https://doc.cgal.org/4.7/Manual/preliminaries.html#seccgal_version).

This PR allows to take into account pre-public internal releases of CGAL-4.7.0

Comment (metux) [reply src/algorithm/Intersection3D.cpp]: I'm using the "releases/CGAL-4.7" tag - sounds like a official release

Comment (lrineau) [reply src/algorithm/Intersection3D.cpp]: OK, now I understand. The tags of our Git repositories are not exactly the releases. CGAL-4.7.0 is the tarball https://github.com/CGAL/cgal/releases/download/releases%2FCGAL-4.7/CGAL-4.7.tar.xz (or its `.zip` equivalent).

Comment (metux) [reply src/algorithm/Intersection3D.cpp]: Uh, the trees look *COMPLETELY* different.
14794 files changed, 3603 insertions(+), 6951836 deletions(-)

This makes things *really* hard for downstreams.

Comment (lrineau) [reply src/algorithm/Intersection3D.cpp]: That are not *that* different. The release tarballs are flattened, compared to the Git repository: the content of all CGAL packages (subdirectories of the root of the Git repository) are copied to a unique directory, and then:
  - files related to the testsuite and documentation are removed,
  - the content of the files are processed to modify the keywords `$Id$` and `$URL$` at the top of all headers files,
  - and other few adjustments (such as in `<CGAL/version.h>`).

For CGAL-5.0 and later, the script producing the release tarballs was the CMake script 
[`/cgal_create_release_with_cmake.cmake`]. For older releases, it was the perl script [`create_internal_release`].

[`/cgal_create_release_with_cmake.cmake`]: https://github.com/CGAL/cgal/blob/releases/CGAL-5.0.1/Scripts/developer_scripts/cgal_create_release_with_cmake.cmake

[`create_internal_release`]: https://github.com/CGAL/cgal/blob/releases/CGAL-5.0.1/Scripts/developer_scripts/create_internal_release

Comment (metux) [reply src/algorithm/Intersection3D.cpp]: > That are not _that_ different. The release tarballs are flattened, compared to the Git repository: the content of all CGAL packages (subdirectories of the root of the Git repository) are copied to a unique directory, and then:
> 
> * files related to the testsuite and documentation are removed,
> * the content of the files are processed to modify the keywords `$Id$` and `$URL$` at the top of all headers files,
> * and other few adjustments (such as in `<CGAL/version.h>`).

You describe exactly that they're *NOT* the same, but the release tarball is a build product of the actual source tree.

Those things are just making life much harder for distro maintainers (note: patching is the regular case for us), as we essentially have to do the inverse transformation for rebasing or upstreaming. Or just ignore the release tarball and only work on git trees.
(Actually, I'm usually working only on git repos, tarballs are the exception for corner cases.)

By the way: why doing that complex release transformation in the first place ?

Comment (lrineau) [reply src/algorithm/Intersection3D.cpp]: > You describe exactly that they're _NOT_ the same, but the release tarball is a build product of the actual source tree.

Yes, they are not the same, but they are not that different either.

> Those things are just making life much harder for distro maintainers (note: patching is the regular case for us), as we essentially have to do the inverse transformation for rebasing or upstreaming. Or just ignore the release tarball and only work on git trees.

Not really, that is actually simple: if the patch applies to the Git repo with `-p1` then it applies to the tarballs with `-p2`. Because:
  - downstream users do not patch anything in the copyright/license headers (that is probably even forbidden by the CGAL license).
  - downstream users usually do not patch the documentation or the test files,
  - and nobody should patch the numbers in `<CGAL/version.h>` but me (I am the release manager of CGAL).

> (Actually, I'm usually working only on git repos, tarballs are the exception for corner cases.)

I understand, we are in 2020. But 10 years ago, that was the reverse: nobody used git repos, because git was rather new, git-submodule was experimental, ang google `repo` was not yet published. 

> By the way: why doing that complex release transformation in the first place ?

Because CGAL is organized in "packages", that historically were authored by different research groups in the UE. Packages were organized as sub-directories, but each of them contributes headers. I can even say that the very first versions of CGAL (more than 20 years ago) were handled by CVS, with different version control repo for the various packages.

In a way, the CGAL version control is organized similarly to the one of the Boost C++ templates libraries (www.boost.org), and probably for the same reasons.


When we switched to Github in 2015, we worked hard so that CGAL sources can be used either from the release tarballs, or from the Git repository. It works so well that you even did not saw, until now, that there was something strange. CGAL version 4.7 was actually the second release (after version 4.6) that was created from a public repo on Github. For all the previous versions, the release were only published as tarballs, and not as git tags.

Comment (metux) [reply src/algorithm/Intersection3D.cpp]: > > You describe exactly that they're _NOT_ the same, but the release tarball is a build product of the actual source tree.
> 
> Yes, they are not the same, but they are not that different either.

For SCM, it matters a lot. We need patches/commits to be easily reviewable, rebase, etc, etc.
At certain point, git's rename detection won't be clever enough anymore, so lots of manual intervention becomes necesary :(
 
> Not really, that is actually simple: if the patch applies to the Git repo with `-p1` then it applies to the tarballs with `-p2`. Because:

So, you seriously expect us to have special workflows for individual packages ?
Sorry, but this adds lots of unncessary extra burden. Please keep in mind that we're
a) usually not active developers of individual libs/applications,
b) maintain lots of different packages,
c) have to pull lots of things together (various combinations of libs on various dists/targets, etc)

> * downstream users usually do not patch the documentation or the test files,

Sometimes we have to do that, at least when it comes to build / deployment.

> I understand, we are in 2020. But 10 years ago, that was the reverse: nobody used git 

I've completely moved to git back 15 years ago. (not even working text-based patches - except for communication @lkml)

> repos, because git was rather new, git-submodule was experimental, ang google `repo` was 
> not yet published.

Actually, I only see few really practical use cases for submodules (and this adds extra trouble on intfrastructure side) - most of the times upstreams uses them, it's just due to pretty bad architecture (lack of proper modularization and release engineering, etc). gnulib is an nice example for such kind of abuse :p

Anyways, cgal doesn't seem to use submodules.

> > By the way: why doing that complex release transformation in the first place ?
> 
> Because CGAL is organized in "packages", that historically were authored by different research groups in the UE. Packages were organized as sub-directories, but each of them contributes headers. 

Okay, and what's the problem here ?

Why not just leaving them as completely different packages ? This would be even much easier for build/deployment - one only needs to install what's really needed, instead of one huge package with tons of dependencies.

> In a way, the CGAL version control is organized similarly to the one of the Boost C++ templates libraries ([www.boost.org](http://www.boost.org)), and probably for the same reasons.

Boost is a real nightmare for dist maintainers. Please don't repeat their fundamental mistakes :p 

--mtx

Comment (lrineau) [reply src/algorithm/Intersection3D.cpp]: > > > You describe exactly that they're _NOT_ the same, but the release tarball is a build product of the actual source tree.
> > 
> > 
> > Yes, they are not the same, but they are not that different either.
> 
> For SCM, it matters a lot. We need patches/commits to be easily reviewable, rebase, etc, etc.
> At certain point, git's rename detection won't be clever enough anymore, so lots of manual intervention becomes necesary :(

As you are currently using the Git repository of CGAL as source distribution, I do not really what you complain about. I can assure you we will handle the situation once you have anything to contribute to CGAL.

> > Not really, that is actually simple: if the patch applies to the Git repo with `-p1` then it applies to the tarballs with `-p2`. Because:
> 
> So, you seriously expect us to have special workflows for individual packages ?

No. I do not expect anything from you. And I do not understand why you complain. As I said earlier, I am the release manager of CGAL, and I do not like the current situation of our source code distribution either. I was just kindly answering your question about the way release tarballs are created. It was not a justification of the architecture.

Since I join the CGAL project, I have triggered a lot of transformations in CGAL. I am the one who worked on permitting the use of the Git repository layout as it is, without the need to use a release tarball. I am also the one who triggered the transition from Subversion to Git, and then later to Github. And there are still a lot of items on my TODO-list. The CGAL project has a lot of history (it is almost as old as the Linux kernel), and it carries the burden of that history. We have worked on cleaning our stuff, and we still do.

> Sorry, but this adds lots of unncessary extra burden. Please keep in mind that we're
> a) usually not active developers of individual libs/applications,
> b) maintain lots of different packages,
> c) have to pull lots of things together (various combinations of libs on various dists/targets, etc)

And, concretely, what are your current issues with CGAL as it is?

As far as package managers are concern, the CGAL project is well aware, and participate in several of them. The packages for Debian/Ubuntu, Fedora, RHEL/CentOS, Homebrew, Conda all use the releases tarballs, but the packaging of CGAL in vcpkg uses a Git tag. And that works. As a package manager, you have the choice.

If you maintain a lot of different packages, you should be use to the fact that the sources as they are in the SCM are not always the sources that are delivered to the public. For any software relying on autoconf, for example, the sources have to be "bootstrapped"/"autoreconfed"... And I know that GNU software are far from being packager-friendly (we had recently to carry a patch to home-made m4 macros in a downstream CGAL project and it was nigthmarish).

Right, that it not convenient for users, or for downstream projects or downstream packagers, and we know it, and we are still working on that. 

> > * downstream users usually do not patch the documentation or the test files,
> 
> Sometimes we have to do that, at least when it comes to build / deployment.

Good. More than 2500 pull-requests were merged in CGAL in five years. We will gladly welcome your contributions.

> > I understand, we are in 2020. But 10 years ago, that was the reverse: nobody used git
> 
> I've completely moved to git back 15 years ago. (not even working text-based patches - except for communication @lkml)

I should have know: you work on the Linux kernel.

> > Because CGAL is organized in "packages", that historically were authored by different research groups in the UE. Packages were organized as sub-directories, but each of them contributes headers.
> 
> Okay, and what's the problem here ?
> 
> Why not just leaving them as completely different packages ? This would be even much easier for build/deployment - one only needs to install what's really needed, instead of one huge package with tons of dependencies.

The problem is testing. We have a plan to modularize CGAL and distribute in distinct packages. But before that we will have to change a lot of our testing infra, because currently CGAL packages are all tested together, from the tree of one Git commit. To modularize them, we will have first to deal with the dependency graph of all packages. [Here is one of my attempt, have fun with it!](https://cgal.geometryfactory.com/~lrineau/CGAL-packages-dependencies/deps.html).

> > In a way, the CGAL version control is organized similarly to the one of the Boost C++ templates libraries ([www.boost.org](http://www.boost.org)), and probably for the same reasons.
> 
> Boost is a real nightmare for dist maintainers. Please don't repeat their fundamental mistakes :p

You are right. It was not a justification. Just a statement that we are not the only open source project with such a bad design, and probably because of the same historical reasons.

Comment (metux) [reply src/algorithm/Intersection3D.cpp]: > As you are currently using the Git repository of CGAL as source distribution, I do not really what you complain about. 

The huge confusion. Usually, I try to stay close to upstream releases ...

> Since I join the CGAL project, I have triggered a lot of transformations in CGAL. I am the one who worked on permitting the use of the Git repository layout as it is, without the need to use a release tarball. I am also the one who triggered the transition from Subversion to Git, and then later to Github. 

Sorry, didn't mean to offend you. And: thanks for all your work.

> And, concretely, what are your current issues with CGAL as it is?

For CGAL:
* huge confusion between git tags and tarballs, including version numbers - any other dist maintainer will run into the same problem and needs time to clear that out
* it's a big thing with tons of dependencies. so we (packages) have to choose between either drop lots of features or create *huge* dependency stack (even to monstreaus gui stuff) for applications that only need a few modules.
* broken crosscompile

For SFCGAL: (see pull requests)
* build system broken in various places
  --> broken library and path lookups
  --> broken installation pathes
* various compatibility issues w/ CGAL 
  --> weird overloading of CGAL headers
  --> broken includes
  --> broken version specific tweaks
  --> type mismatches
  --> using _internal_ CGAL data structures
  --> missing pkg-config
  --> broken crosscompile

> For any software relying on autoconf, for example, the sources have to be "bootstrapped"/"autoreconfed"... 

I know, and we've got well established procedures for that (eg. dh_autoconf etc),
it's a standard build stage, just like configure or make.

autotools world used to be real nightmare back 20 years ago, but became much better,
especially once we got libtool repaired and pkg-config became a standard tool.

> And I know that GNU software are far from being packager-friendly (we had recently to carry a patch to home-made m4 macros in a downstream CGAL project and it was nigthmarish).

Talking about stuff directly from the foundation ? Yes, there's a lot crap there, gnulib for example (I wouldn't even call this thing a 'package'). :p 

> Good. More than 2500 pull-requests were merged in CGAL in five years. We will gladly welcome your contributions.

Thanks :)

> I should have know: you work on the Linux kernel.

Yes. I'm a maintainer :p 

> > Why not just leaving them as completely different packages ? This would be even much easier for build/deployment - one only needs to install what's really needed, instead of one huge package with tons of dependencies.
> 
> The problem is testing. We have a plan to modularize CGAL and distribute in distinct packages. But before that we will have to change a lot of our testing infra, because currently CGAL packages are all tested together, from the tree of one Git commit.

Setting up an automatic subtree-merge shouldn't be a big deal. Existing CI would then act on the automerged tree.

OTOH, I believe it would be a good thing to add building dist packages and use them for further build steps (eg. let the sfcgal jobs just install cgal binary packages instead of compiling on its own).

> To modularize them, we will have first to deal with the dependency graph of all packages. 
> [Here is one of my attempt, have fun with it!]
> (https://cgal.geometryfactory.com/~lrineau/CGAL-packages-dependencies/deps.html).

Maybe I find some time to drop out certain things that I don't need for postgis (or at least drop out what sfcgal doesn't need). This should give us a better idea how to do this practically.

--mtx
